From db2bb3f92b43c58d68a521bbb21f653ad36006e4 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Mon, 11 Sep 2023 21:37:29 -0500
Subject: [PATCH 1/2] same as alwayscenter patch

but allows the user to center floating relative to monitor or window area
---
 config.def.h | 1 +
 dwl.c        | 6 +++++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/config.def.h b/config.def.h
index 1677f6f12..7b473575c 100644
--- a/config.def.h
+++ b/config.def.h
@@ -6,6 +6,7 @@ static const float bordercolor[]           = {0.5, 0.5, 0.5, 1.0};
 static const float focuscolor[]            = {1.0, 0.0, 0.0, 1.0};
 /* To conform the xdg-protocol, set the alpha to zero to restore the old behavior */
 static const float fullscreen_bg[]         = {0.1, 0.1, 0.1, 1.0};
+static const int center_relative_to_monitor = 0; /* 0 means center floating relative to the window area */
 
 /* tagging - tagcount must be no greater than 31 */
 #define TAGCOUNT (9)
diff --git a/dwl.c b/dwl.c
index 4ff5c37f7..c7cc95cc0 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1536,7 +1536,7 @@ mapnotify(struct wl_listener *listener, void *data)
 {
 	/* Called when the surface is mapped, or ready to display on-screen. */
 	Client *p, *w, *c = wl_container_of(listener, c, map);
-	Monitor *m;
+	Monitor *m = selmon;
 	int i;
 
 	/* Create scene tree for this client and its border */
@@ -1582,6 +1582,10 @@ mapnotify(struct wl_listener *listener, void *data)
 	wl_list_insert(&clients, &c->link);
 	wl_list_insert(&fstack, &c->flink);
 
+	struct wlr_box b = center_relative_to_monitor ? m->m : m->w; 
+	c->geom.x = (b.width - c->geom.width) / 2 + b.x;
+	c->geom.y = (b.height - c->geom.height) / 2 + b.y;
+
 	/* Set initial monitor, tags, floating status, and focus:
 	 * we always consider floating, clients that have parent and thus
 	 * we set the same tags and monitor than its parent, if not

From 99a9016289f75aaaf17e53459883f7ee8414e703 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Mon, 11 Sep 2023 22:26:53 -0500
Subject: [PATCH 2/2] fix warning

---
 dwl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dwl.c b/dwl.c
index c7cc95cc0..8ef4abdf3 100644
--- a/dwl.c
+++ b/dwl.c
@@ -1538,6 +1538,7 @@ mapnotify(struct wl_listener *listener, void *data)
 	Client *p, *w, *c = wl_container_of(listener, c, map);
 	Monitor *m = selmon;
 	int i;
+	struct wlr_box b = center_relative_to_monitor ? m->m : m->w; 
 
 	/* Create scene tree for this client and its border */
 	c->scene = wlr_scene_tree_create(layers[LyrTile]);
@@ -1582,7 +1583,6 @@ mapnotify(struct wl_listener *listener, void *data)
 	wl_list_insert(&clients, &c->link);
 	wl_list_insert(&fstack, &c->flink);
 
-	struct wlr_box b = center_relative_to_monitor ? m->m : m->w; 
 	c->geom.x = (b.width - c->geom.width) / 2 + b.x;
 	c->geom.y = (b.height - c->geom.height) / 2 + b.y;
 
